File: ruby/zipper/zipper.rb
# Thanks to @mengqing for the inspiration.

############################################################
class Node 
  def initialize(value, left, right)
    @value = value
    @left  = left
    @right = right
  end
  attr_accessor :value, :left, :right

  def ==(other)
    other.class == self.class && other.state == state
  end

  protected

  def state
    [value, left, right]
  end
end

############################################################
class Zipper
  def self.from_tree(tree)
    new(tree)
  end

  def initialize(tree, path = [])
    @tree = tree
    @path = path
  end
  attr_reader :tree

  def to_tree
    @path.first&.tree || @tree
  end

  def left
    return if @tree.left.nil?
    self.class.new(@tree.left, @path + [self])
  end

  def right
    return if @tree.right.nil?
    self.class.new(@tree.right, @path + [self])
  end

  def value
    @tree.value
  end

  def up
    return if @path.empty?
    previous = @path.pop
    self.class.new(previous.tree, @path)
  end

  # rubocop:disable Naming/AccessorMethodName

  def set_value(value)
    @tree.value = value
    self
  end

  def set_left(tree)
    @tree.left = tree
    self
  end

  def set_right(tree)
    @tree.right = tree
    self
  end

  # rubocop:enable Naming/AccessorMethodName

  def ==(other)
    other.class == self.class && other.state == state
  end

  protected

  def state
    [tree]
  end
end

File: javascript/zipper/zipper.js
const deepCopy = obj => JSON.parse(JSON.stringify(obj));

export class Zipper {
  constructor(tree, path = []) {
    this.tree = tree;
    this.path = path;
  }

  static fromTree(tree) { return new Zipper(deepCopy(tree)); }

  getTree() { return this.tree; }

  toTree() {
    return this.path.length > 0
      ? this.path[0].getTree()
      : this.tree;
  }

  value() { return this.tree.value; }

  left() {
    return this.tree.left
      ? new Zipper(this.tree.left, this.path.concat(this))
      : null;
  }

  right() {
    return this.tree.right
      ? new Zipper(this.tree.right, this.path.concat(this))
      : null;
  }

  up() {
    return this.path.length > 0
      ? new Zipper(this.path.pop().getTree(), this.path)
      : null;
  }

  setValue(value) {
    this.tree.value = value;
    return new Zipper(this.toTree());
  }

  setLeft(tree) {
    this.tree.left = tree;
    return new Zipper(this.toTree());
  }

  setRight(tree) {
    this.tree.right = tree;
    return new Zipper(this.toTree());
  }
}
